### PR TITLE
Refactoring: FAT-82 & 99 Onboarding polling hook refactor

### PR DESCRIPTION
**Pull request by alexandremgo**
### 📝 Description

Refactoring of the useOnboardingStatePolling that has no impact on current develop/production projects. This PR merged on `develop` will make our work easier on both LLD and LLM.

This refactoring brings 2 changes:
- avoid re-rendering: the hook only updates its result on a new onboardingState or new allowedError, not at every run
- update of `useOnboardingStatePolling args`: `getOnboardingStatePolling` as an optional injected dependency to the hook. It is needed for LLD to have the polling working on the **internal thread**. It is set by default to `live-common/hw/getOnboardingStatePolling` so it is not needed to pass it as an arg to use the hook on LLM.

Note: you can see a usage of the "dependency injection"/LLD on commit [371ada8c5f7c8d3cf63f9974d1756933d671e21a](https://github.com/LedgerHQ/ledger-live/pull/770/commits/371ada8c5f7c8d3cf63f9974d1756933d671e21a).

### ❓ Context

- **Impacted projects**: live-common
- **Linked resource(s)**: Will be used by [FAT-82 on LLM](https://ledgerhq.atlassian.net/browse/FAT-82?atlOrigin=eyJpIjoiMzllODNlOTNmYTVjNDdiZTkwZDUxYjkyYWU0ZWZkNGYiLCJwIjoiaiJ9) and [FAT-99 on LLD](https://ledgerhq.atlassian.net/browse/FAT-99?atlOrigin=eyJpIjoiYWM1NTc0YTVlMmVkNGIxOTgyMzIxNWYyN2M2NGY3Y2EiLCJwIjoiaiJ9)

### ✅ Checklist

- [x] **Test coverage** : unit tested
- [x] **Atomic delivery** 
- [x] **No breaking changes** : no impact on current develop/production projects

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

File: libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts [hunk 1]
```diff
@@ -80,7 +120,26 @@ export const useOnboardingStatePolling = ({
...
+
+const isDifferentOnboardingState = (
+  prevOnboardingState: OnboardingState | null,
+  newOnboardingState: OnboardingState | null
+): boolean => {
+  // OnboardingState are simple JSON-style objects, without methods
+  return (
+    JSON.stringify(prevOnboardingState) !== JSON.stringify(newOnboardingState)
+  );
+};
```
Comment (mrbaguvix) [inline libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts:137]: Looks good. Another interesting thing i suggest is this same function can handle the entire logic and just return` onboardingStatePollingResult.onboardingState`

```
if(....){
return newOnboardingState
}

return prevOnboardingState
```

File: libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts [hunk 2]
```diff
@@ -49,11 +65,35 @@ export const useOnboardingStatePolling = ({
...
+                if (
+                  isDifferentOnboardingState(
+                    prevOnboardingState,
+                    onboardingStatePollingResult.onboardingState
+                  )
+                ) {
+                  return onboardingStatePollingResult.onboardingState;
+                }
+
+                return prevOnboardingState;
```
Comment (mrbaguvix) [inline libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts:95]: We can simplify the state function by moving all the logic to `isDifferentOnboardingState`
`return isDifferentOnboardingState(...)`

File: libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts [hunk 3]
```diff
@@ -49,11 +65,35 @@ export const useOnboardingStatePolling = ({
...
+              if (
+                isDifferentError(
+                  prevAllowedError,
+                  onboardingStatePollingResult.allowedError
+                )
+              ) {
+                return onboardingStatePollingResult.allowedError;
+              }
+
+              return prevAllowedError;
```
Comment (mrbaguvix) [inline libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts:79]: I suggest the logic be moved into `isDifferentError` function

File: libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts [hunk 4]
```diff
@@ -80,7 +120,26 @@ export const useOnboardingStatePolling = ({
...
+    JSON.stringify(prevOnboardingState) !== JSON.stringify(newOnboardingState)
+  );
+};
+
+const isDifferentError = (
+  prevError: Error | null,
+  newError: Error | null
+): boolean => {
+  // Only interested if the errors are instances of the same Error class
+  return prevError?.constructor !== newError?.constructor;
```
Comment (mrbaguvix) [inline libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts:144]: Can we try 
```js 
prevError instanceof newError
```

File: libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts (hunk 2 above)
Comment (alexandremgo) [reply libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts:95]: Hi @mrbaguvix ! Thanks for your review :) I agree, i'll simplify with the state function by moving the logic in one function

File: libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts (hunk 4 above)
Comment (alexandremgo) [reply libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts:144]: `instanceof` needs a "Constructor" on the right side ([doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof#syntax)). So `prevError instanceof newError` would not work. 

The easiest in my case is just to check the constructor

**Review by juan-cortes — APPROVED** [review]

**Pull request merged**